lib/mix/lib/mix/tasks/compile.xref.ex

```diff
@@ -28,28 +29,34 @@ defmodule Mix.Tasks.Compile.Xref do
     {opts, _, _} =
       OptionParser.parse(args, switches: [force: :boolean, warnings_as_errors: :boolean])
 
-    if needs_xref?(opts) do
-      if should_exit?(run_xref(), opts) do
-        exit({:shutdown, 1})
+    warnings =
+      if needs_xref?(opts) do
+        run_xref()
+      else
+        read_manifest()
       end
-      write_manifest()
-    end
-
-    :noop
-  end
 
-  defp run_xref do
-    Mix.Task.run("xref", ["warnings"])
+    if warnings != [] and warnings_as_errors(opts) do
+      {:error, to_diagnostics(warnings, :error)}
+    else
+      {:noop, to_diagnostics(warnings, :warning)}
+    end
   end
 
   defp needs_xref?(opts) do
     !!opts[:force] or Mix.Utils.stale?(E.manifests(), manifests())
   end
 
-  defp should_exit?(:error, opts),
-    do: warnings_as_errors(opts) == true
-  defp should_exit?(_, _opts),
-    do: false
+  defp run_xref do
+    timestamp = :calendar.universal_time()
+    warnings = Mix.Task.run("xref", ["warnings"])
```

Comment:
Let's make it return `{:ok, warnings}` (and then `:noop` if it already ran) so it forces us to properly pattern match.